app/src/main/java/com/epfl/drawyourpath/database/FirebaseDatabase.kt

```diff
@@ -463,6 +488,86 @@ class FirebaseDatabase : Database() {
         return Utils.failedFuture(Error("Not implemented"))
     }
 
+    override fun getTournamentUID(): String? {
+        return database.child(FirebaseKeys.TOURNAMENTS_ROOT).push().key
+    }
+
+    override fun addTournament(tournament: Tournament): CompletableFuture<Unit> {
+        val future = CompletableFuture<Unit>()
+        // add the tournament to all tournaments
+        tournamentsRoot().updateChildren(mapOf(tournament.id to tournament))
+            .addOnSuccessListener { future.complete(Unit) }
+            .addOnFailureListener { future.completeExceptionally(it) }
+        return future
+    }
+
+    override fun removeTournament(tournamentId: String): CompletableFuture<Unit> {
+        val future = CompletableFuture<Unit>()
+
+        database.child("${FirebaseKeys.TOURNAMENTS_ROOT}/$tournamentId/${FirebaseKeys.TOURNAMENT_PARTICIPANTS_IDS}")
+            .get()
+            .addOnSuccessListener { data ->
+                // get the list of all participants (list of ids)
+                val participantsIds = transformSnapshotKeysToStringList(data)
+                // prepare changes in database
+                // 1. remove the tournament from the list of tournaments of all participants
+                val changes: MutableMap<String, Any?> = participantsIds.associate {
+                    "${FirebaseKeys.USERS_ROOT}/$it/${FirebaseKeys.USER_TOURNAMENTS}/$tournamentId" to null
+                }.toMutableMap()
+                // 2. remove the tournament from the tournaments file
+                changes.put("${FirebaseKeys.TOURNAMENTS_ROOT}/$tournamentId", null)
+                // do the changes
+                database.updateChildren(changes).addOnSuccessListener { future.complete(Unit) }
+                    .addOnFailureListener { future.completeExceptionally(it) }
+            }
+            .addOnFailureListener { future.completeExceptionally(it) }
+
+        return future
+    }
```

Comment:
I don't think we can make this code work with security rules.
Also, if we have let's say a million participant registered to a tournament, a client will make a million request emptying his mobile data. It doesn't really scale.

I think we could keep it like that for this sprint, but we should only mark a tournament as deleted and keep it in the db. We probably are going to want to show a list of past tournaments. What do you think?

Reply:
I will be honest I have no idea about the security rules, but I guess that a user deleting a tournament should not change the user file of another user? If it's true then this function could be reserved to "admins"... I don't really know.

Indeed it is a problem that it doesn't scale well and I did not think about it. However I would argue that it is still useful to have such a function to be able to really remove a tournament. Maybe, as I said above, we could only use it in really special situations, when we don't have much choice (but only as admins). 

We certainly should not let users delete tournaments with this function all the times if there are large numbers of participants, and we should be careful when deleting weekly tournament. 

I can see two solutions for now:
- limit the number of participants in tournaments that are not weekly (but requires to distinguish the weekly tournaments).
- "never" delete tournaments as you said, except if the tournament has not started for example.

Unfortunately I don't have time to change all of this for this sprint, but we should talk about that with the whole team on a meeting, to see what should be done.

I hope everything I said makes sense, let me know what you think.

Reply:
I think the second option is best. We can talk about it in the next meeting and plan it for the next sprint.